### PR TITLE
.itfirst bit of blob handling

### PR DESCRIPTION
**Pull request by mredig**
I had a lot of trouble with the way that reading chars into ints changes (from memory vs file handle) depending on the context (return value vs intvalue = charvalue with being signed vs not and etc), but I finally got it to work with completely copying the flif file into memory and then exclusively decoding from it (at least from what I can tell in the very hard to read code). It still can only write back out to a file handle yet, but I want to further extend it to retrieve the raw image and be able to return it to an external function.

Also, I'm not sure how to selectively submit commits for a pull request, so the previous denied commits are here as well. Maybe you can cull them in the merge process?

### CODE REVIEW

**Comment by mredig** [comment]
Good catch. I'll have to update this when I get a chance.

File: fileio.h [hunk 1]
```diff
@@ -51,4 +51,151 @@ class FileIO
...
+};
+
+
+class FlifBlobIO
+{
+private:
+    FILE *file;
+    const char *name;
+    
+    char* flifBlob;
```
Comment (sipa) [inline fileio.h]: Can you use a C++ data type here that offers automatic memory freeing (std::vector<char> would work)?

File: fileio.h [hunk 2]
```diff
@@ -51,4 +51,151 @@ class FileIO
...
+    void operator=(const FlifBlobIO&) {}
+    // prevent move, for now
+    FlifBlobIO(FlifBlobIO&&) {}
+    void operator=(FlifBlobIO&&) {}
+public:
+    FlifBlobIO(FILE* fil, const char *aname){
+        readFlifFromFile(fil, aname);
+    }
+    
+    FlifBlobIO(FileIO *fio) {
```
Comment (sipa) [inline fileio.h]: There should be no need to convert a file into a blob. If you have a file, use FileIO. If you have a blob, use BlobIO.

File: fileio.h [hunk 3]
```diff
@@ -51,4 +51,151 @@ class FileIO
...
       return name;
     }
+    
+    FILE* getFileHandle() {
+        return file;
+    }
+};
+
+
+class FlifBlobIO
```
Comment (sipa) [inline fileio.h]: Nothing about this class should be specific to FLIF,

File: flif-dec.cpp [hunk 4]
```diff
@@ -277,13 +277,13 @@ template <typename IO>
 bool flif_decode(IO& io, Images &images, int quality, int scale)
 {
     if (scale != 1 && scale != 2 && scale != 4 && scale != 8 && scale != 16 && scale != 32 && scale != 64 && scale != 128) {
-                e_printf("Invalid scale down factor: %i\n", scale);
-                return false;
+        e_printf("Invalid scale down factor: %i\n", scale);
```
Comment (sipa) [inline flif-dec.cpp]: Can you avoid unrelated changes?

File: flif.cpp [hunk 5]
```diff
@@ -216,8 +227,9 @@ int handle_decode_arguments(char **argv, Images &images, int quality, int scale)
     if(!file)
         return 1;
     FileIO fio(file, argv[0]);
+    FlifBlobIO bio(&fio);
```
Comment (sipa) [inline flif.cpp]: This means you're going to load the entire file in memory immediately, plus break support for streaming decode. You should just use FileIO when a file is passed (which will always be the case in this function, AFAICT), and use flif_decode on a BlobIO whenever you have a blob only.

File: fileio.h (hunk 1 above)
Comment (mredig) [inline fileio.h]: So it would be a vector containing chars?

Would that potentially cause issues when interfacing with other host programming languages? Or is there a way to convert char* to a vector and back, easily (and ideally without doubling memory consumption)?

File: fileio.h (hunk 2 above)
Comment (mredig) [inline fileio.h]: I did this simply for testing purposes as there is no way to get a blob as of yet. Convenience in testing is all.

File: flif-dec.cpp (hunk 4 above)
Comment (mredig) [inline flif-dec.cpp]: I'm not sure how to break pull requests into separate pull requests - GitHub seem so just update a single pull request when I add commits.

File: flif.cpp (hunk 5 above)
Comment (mredig) [inline flif.cpp]: Testing purposes again. 

I suppose it would be wise to add comments to lines like these.

File: fileio.h [hunk 6]
```diff
@@ -54,4 +55,206 @@ class FileIO
...
+    // prevent move, for now
+    BlobIO(BlobIO&&) {}
+    void operator=(BlobIO&&) {}
+public:
+    BlobIO(FILE* fil, const char *aname){
+        readFlifFromFile(fil, aname);
+		closeFile();
+    }
+    
+    BlobIO(FileIO *fio) { //used only for testing - will remove when there's a way to get blobs elsewhere
```
Comment (jonsneyers) [inline fileio.h]: Testing code should not be here. Best to remove it or put it in a separate file (which you don't need to commit).

File: fileio.h [hunk 7]
```diff
@@ -54,4 +55,206 @@ class FileIO
...
+        }
+        initializeVars();
+    }
+	
+	void closeFile() {
+		fclose(file);
+		file = NULL;
+	}
+	
+    void setBlob(char* blobFlif, size_t size) {
```
Comment (jonsneyers) [inline fileio.h]: Can the copy be avoided somehow?

File: fileio.h [hunk 8]
```diff
@@ -54,4 +55,206 @@ class FileIO
...
+    size_t getFlifSize() {
+        return blob.size();
+    }
+    
+    const char* getCharBlob() {
+        char* flifBlob = new char[blob.size()];
+        for (size_t i = 0; i < blob.size(); i++) {
+            flifBlob[i] = blob[i];
+        }
+        return flifBlob;
```
Comment (jonsneyers) [inline fileio.h]: I think `return &blob[0]` should work.

File: fileio.h [hunk 9]
```diff
@@ -54,4 +55,206 @@ class FileIO
...
+                break;
+            case SEEK_SET:
+                memoryPos = offset;
+                break;
+                
+            default:
+                break;
+        }
+        if (memoryPos > fSize) {
+            memoryPos = fSize-1;
```
Comment (jonsneyers) [inline fileio.h]: Wouldn't this put seeks beyond the end put the position at (last byte) instead of (EOF) ?

File: fileio.h [hunk 10]
```diff
@@ -54,4 +55,206 @@ class FileIO
...
+    }
+    
+    void write(int byte) {
+        fputc(byte);
+    }
+    void flush() {
+//        fflush(file); //// I don't think this would really have an equivalent for writing to memory, since it writes directly to memory in the first place?
+    }
+    bool isEOF() {
+        return (memoryPos >= blob.size() ); //not sure if >= or >
```
Comment (jonsneyers) [inline fileio.h]: `>=` is right. `blob[blob.size()]` is an out-of-bounds index.

File: fileio.h (hunk 7 above)
Comment (mredig) [inline fileio.h]: I specifically want to be able to share blobs with Obj C bidirectionally. Since the class uses vectors internally, I don't believe there's any way to pass information between the two languages without generic c types.

File: fileio.h (hunk 8 above)
Comment (mredig) [inline fileio.h]: The returned pointer could become invalid if there's a reallocation of the internal array.

File: fileio.h (hunk 9 above)
Comment (mredig) [inline fileio.h]: I'm not sure if EOF is n-1 or n; Is it 0 indexed or not? If 0, EOF should theoretically be n-1, otherwise you're right. If you or someone else doesn't know for sure, I can do some experimentation in the next couple days.

File: fileio.h (hunk 7 above)
Comment (sipa) [inline fileio.h]: It can call class methods, so it understands classes... but can only pass pure C types to it? That seems weird.

I would suggest to keep the internal classes clean and not confined to what an external API demands. You'd then create some wrapper classes/functions in C++ code which are callable by Obj C, and only compiled when building those bindings.

Comment (mredig) [inline fileio.h]: I don't honestly know. I'd have to experiment to be certain, but blob handling is not quite to where I can make the appropriate experiment yet.

I think that suggestion would be fine for ME (assuming I can't pass it directly), but IMO that makes it more confusing to other developers down the road who may not see that as a possibility.

File: fileio.h (hunk 10 above)
Comment (mredig) [inline fileio.h]: I'm looking at this and realizing, "DUH".

Thanks for reminding me! :)

File: fileio.h (hunk 7 above)
Comment (sipa) [inline fileio.h]: Sorry, I disagree. Having a clear abstraction exposed by a wrapper that fits the other language is almost certainly better than trying to expose internal C++ code directly.

File: fileio.h (hunk 6 above)
Comment (mredig) [inline fileio.h]: I don't know how to have a second file as part of the same class. Any hints?

Comment (sipa) [inline fileio.h]: No, you would create a separate wrapper class for the interface. The internal class here wouldn't carry any logic specific to the other language.

Comment (mredig) [inline fileio.h]: Yeah I suppose I could make a separate function, but I meant a separate file that augments the BlobIO class. Is that possible? (essentially a second header file that behaves as if it's part of the regular class)

Comment (sipa) [inline fileio.h]: It is possible to have a method defintion in a class for which the
implementation is in an unrelated .cpp file. That's pretty terrible code
organization though.

What is also possible is friend methods and classes... other code that is
allowed to access private fields.

Not a suggestion to use either of those here, though.

**Comment by jonsneyers** [comment]
Thanks for the efforts! This PR has been made redundant by this one: https://github.com/FLIF-hub/FLIF/pull/110
